OBSOLETE (put tools tests in reporoot/tools/test/): add mechanism to allow for tools imports in our tests

dagitses opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63947
* #64439
* __->__ #63946

We of course want to have lower level tests of our codegen code. But
right now we can't import any of the code to use in a non-end-to-end
manner.

See #61296 which is trying to get a codegen test in CI for more
motivation.

Differential Revision: [D30541486](https://our.internmc.facebook.com/intern/diff/D30541486)

Commit:
add mechanism to allow for tools imports in our tests

We of course want to have lower level tests of our codegen code. But
right now we can't import any of the code to use in a non-end-to-end
manner.

See #61296 which is trying to get a codegen test in CI for more
motivation.

[ghstack-poisoned]

Commit:
Update on "add mechanism to allow for tools imports in our tests"


We of course want to have lower level tests of our codegen code. But
right now we can't import any of the code to use in a non-end-to-end
manner.

See #61296 which is trying to get a codegen test in CI for more
motivation.

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541486).

albanD commented on test/tools_package.py:
```diff
@@ -0,0 +1,54 @@
+"""Executable module that adds makes the pytorch/tools modules available for import.
```
nit extra adds

albanD commented on test/tools_package.py:
```diff
@@ -0,0 +1,54 @@
+"""Executable module that adds makes the pytorch/tools modules available for import.
+
+Why does this exist? We want to be able to write unit tests for the
```
nit: newline after the question mark

albanD commented on test/tools_package.py:
```diff
@@ -0,0 +1,54 @@
+import sys
+
+
+def _import_tools_module():
```
Should this live in `torch/testing`?

albanD commented on test/tools_package.py:
```diff
@@ -0,0 +1,54 @@
+
+    # The tools/ "package" is right below the root.
+    TOOLS_DIR = PYTORCH_DIR / 'tools/'
+    assert TOOLS_DIR.is_dir()
```
I think we want a better error message than just an assert here.
Keeping in mind that this might get raised when someone tries to use this file, it should explain that this is most likely because you're not running this from a clone of pytorch that contains the tools directory and maybe link to this file for more details.

albanD reviewed (commented):
Looks good to me.
My main question is about the location of this file. In particular, I don't think we have utility functions in the test folder right now, they are in `torch/testing` (which is installed alongside torch). So that might be a better place to put this utility function.

Commit:
Update on "add mechanism to allow for tools imports in our tests"


We of course want to have lower level tests of our codegen code. But
right now we can't import any of the code to use in a non-end-to-end
manner.

See #61296 which is trying to get a codegen test in CI for more
motivation.

Differential Revision: [D30541486](https://our.internmc.facebook.com/intern/diff/D30541486)

[ghstack-poisoned]

Commit:
Update on "add mechanism to allow for tools imports in our tests"


We of course want to have lower level tests of our codegen code. But
right now we can't import any of the code to use in a non-end-to-end
manner.

See #61296 which is trying to get a codegen test in CI for more
motivation.

Differential Revision: [D30541486](https://our.internmc.facebook.com/intern/diff/D30541486)

[ghstack-poisoned]

Commit:
Update on "add mechanism to allow for tools imports in our tests"


We of course want to have lower level tests of our codegen code. But
right now we can't import any of the code to use in a non-end-to-end
manner.

See #61296 which is trying to get a codegen test in CI for more
motivation.

Differential Revision: [D30541486](https://our.internmc.facebook.com/intern/diff/D30541486)

[ghstack-poisoned]

Commit:
Update on "add mechanism to allow for tools imports in our tests"


We of course want to have lower level tests of our codegen code. But
right now we can't import any of the code to use in a non-end-to-end
manner.

See #61296 which is trying to get a codegen test in CI for more
motivation.

Differential Revision: [D30541486](https://our.internmc.facebook.com/intern/diff/D30541486)

[ghstack-poisoned]

albanD reviewed (approved):
SGTM

Commit:
Update on "add mechanism to allow for tools imports in our tests"


We of course want to have lower level tests of our codegen code. But
right now we can't import any of the code to use in a non-end-to-end
manner.

See #61296 which is trying to get a codegen test in CI for more
motivation.

Differential Revision: [D30541486](https://our.internmc.facebook.com/intern/diff/D30541486)

[ghstack-poisoned]

Commit:
Update on "add mechanism to allow for tools imports in our tests"


We of course want to have lower level tests of our codegen code. But
right now we can't import any of the code to use in a non-end-to-end
manner.

See #61296 which is trying to get a codegen test in CI for more
motivation.

Differential Revision: [D30541486](https://our.internmc.facebook.com/intern/diff/D30541486)

[ghstack-poisoned]

Commit:
Update on "add mechanism to allow for tools imports in our tests"


We of course want to have lower level tests of our codegen code. But
right now we can't import any of the code to use in a non-end-to-end
manner.

See #61296 which is trying to get a codegen test in CI for more
motivation.

Differential Revision: [D30541486](https://our.internmc.facebook.com/intern/diff/D30541486)

[ghstack-poisoned]

Commit:
Update on "add mechanism to allow for tools imports in our tests"


We of course want to have lower level tests of our codegen code. But
right now we can't import any of the code to use in a non-end-to-end
manner.

See #61296 which is trying to get a codegen test in CI for more
motivation.

Differential Revision: [D30541486](https://our.internmc.facebook.com/intern/diff/D30541486)

[ghstack-poisoned]

Commit:
Update on "add mechanism to allow for tools imports in our tests"


We of course want to have lower level tests of our codegen code. But
right now we can't import any of the code to use in a non-end-to-end
manner.

See #61296 which is trying to get a codegen test in CI for more
motivation.

Differential Revision: [D30541486](https://our.internmc.facebook.com/intern/diff/D30541486)

[ghstack-poisoned]

Commit:
Update on "add mechanism to allow for tools imports in our tests"


We of course want to have lower level tests of our codegen code. But
right now we can't import any of the code to use in a non-end-to-end
manner.

See #61296 which is trying to get a codegen test in CI for more
motivation.

Differential Revision: [D30541486](https://our.internmc.facebook.com/intern/diff/D30541486)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541486).

Commit:
Update on "add mechanism to allow for tools imports in our tests"


We of course want to have lower level tests of our codegen code. But
right now we can't import any of the code to use in a non-end-to-end
manner.

See #61296 which is trying to get a codegen test in CI for more
motivation.

Differential Revision: [D30541486](https://our.internmc.facebook.com/intern/diff/D30541486)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541486).

Commit:
Update on "add mechanism to allow for tools imports in our tests"


We of course want to have lower level tests of our codegen code. But
right now we can't import any of the code to use in a non-end-to-end
manner.

See #61296 which is trying to get a codegen test in CI for more
motivation.

Differential Revision: [D30541486](https://our.internmc.facebook.com/intern/diff/D30541486)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541486).

Commit:
Update on "add mechanism to allow for tools imports in our tests"


We of course want to have lower level tests of our codegen code. But
right now we can't import any of the code to use in a non-end-to-end
manner.

See #61296 which is trying to get a codegen test in CI for more
motivation.

Differential Revision: [D30541486](https://our.internmc.facebook.com/intern/diff/D30541486)

[ghstack-poisoned]

Commit:
Update on "add mechanism to allow for tools imports in our tests"


We of course want to have lower level tests of our codegen code. But
right now we can't import any of the code to use in a non-end-to-end
manner.

See #61296 which is trying to get a codegen test in CI for more
motivation.

Differential Revision: [D30541486](https://our.internmc.facebook.com/intern/diff/D30541486)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541486).

Commit:
Update on "add mechanism to allow for tools imports in our tests"


We of course want to have lower level tests of our codegen code. But
right now we can't import any of the code to use in a non-end-to-end
manner.

See #61296 which is trying to get a codegen test in CI for more
motivation.

Differential Revision: [D30541486](https://our.internmc.facebook.com/intern/diff/D30541486)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541486).

Commit:
Update on "add mechanism to allow for tools imports in our tests"


We of course want to have lower level tests of our codegen code. But
right now we can't import any of the code to use in a non-end-to-end
manner.

See #61296 which is trying to get a codegen test in CI for more
motivation.

Differential Revision: [D30541486](https://our.internmc.facebook.com/intern/diff/D30541486)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541486).

Commit:
Update on "add mechanism to allow for tools imports in our tests"


We of course want to have lower level tests of our codegen code. But
right now we can't import any of the code to use in a non-end-to-end
manner.

See #61296 which is trying to get a codegen test in CI for more
motivation.

Differential Revision: [D30541486](https://our.internmc.facebook.com/intern/diff/D30541486)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541486).

Commit:
Update on "add mechanism to allow for tools imports in our tests"


We of course want to have lower level tests of our codegen code. But
right now we can't import any of the code to use in a non-end-to-end
manner.

See #61296 which is trying to get a codegen test in CI for more
motivation.

Differential Revision: [D30541486](https://our.internmc.facebook.com/intern/diff/D30541486)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541486).

dagitses commented:
> SGTM

and in a surprise twist, the test utility has moved back to torch/testing/_internal/ since that ends up working better in OSS and internally.

Commit:
Update on "add mechanism to allow for tools imports in our tests"


We of course want to have lower level tests of our codegen code. But
right now we can't import any of the code to use in a non-end-to-end
manner.

See #61296 which is trying to get a codegen test in CI for more
motivation.

Differential Revision: [D30541486](https://our.internmc.facebook.com/intern/diff/D30541486)

[ghstack-poisoned]

Commit:
Update on "add mechanism to allow for tools imports in our tests"


We of course want to have lower level tests of our codegen code. But
right now we can't import any of the code to use in a non-end-to-end
manner.

See #61296 which is trying to get a codegen test in CI for more
motivation.

Differential Revision: [D30541486](https://our.internmc.facebook.com/intern/diff/D30541486)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541486).

Commit:
Update on "add mechanism to allow for tools imports in our tests"


We of course want to have lower level tests of our codegen code. But
right now we can't import any of the code to use in a non-end-to-end
manner.

See #61296 which is trying to get a codegen test in CI for more
motivation.

Differential Revision: [D30541486](https://our.internmc.facebook.com/intern/diff/D30541486)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541486).

Commit:
Update on "add mechanism to allow for tools imports in our tests"


We of course want to have lower level tests of our codegen code. But
right now we can't import any of the code to use in a non-end-to-end
manner.

See #61296 which is trying to get a codegen test in CI for more
motivation.

Differential Revision: [D30541486](https://our.internmc.facebook.com/intern/diff/D30541486)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541486).

Commit:
Update on "add mechanism to allow for tools imports in our tests"


We of course want to have lower level tests of our codegen code. But
right now we can't import any of the code to use in a non-end-to-end
manner.

See #61296 which is trying to get a codegen test in CI for more
motivation.

Differential Revision: [D30541486](https://our.internmc.facebook.com/intern/diff/D30541486)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30541486).

dagitses commented:
After discussions with @driazati we have come to the conclusion that tests for the codegen are better off living in <reporoot>/tools/test. This directory already exists and is tested by CI.

The downside of this is that test_autograd.py does not test autograd codegen. I think that is acceptable and arguably superior.

I am downgrading to a draft for now, but we can revisit if you have another opinion.
